src/test/java/htsjdk/variant/bcf2/BCF2WriterUnitTest.java

```diff
@@ -0,0 +1,301 @@
+            container =
+                    new VariantContextTestProvider.VariantContextContainer(vcfHeader, new VariantContextTestProvider.VCIterable(codec, vcfHeader) {
+                        @Override
+                        public boolean hasNext() {
+                            try {
+                                return !bodyPbs.isDone();
+                            } catch (IOException e) {
+                                throw new RuntimeException(e);
+                            }
+                        }
+
+                        @Override
+                        public Object nextSource() {
+                            return bodyPbs;
+                        }
+                    });
+
+            int counter = 0;
+            final Iterator<VariantContext> it = container.getVCs().iterator();
+            while (it.hasNext()) {
+                it.next();
+                counter++;
+            }
+            Assert.assertEquals(counter, 2);
+        }
+
+    }
+
+    @Test(expectedExceptions = IllegalStateException.class)
+    public void testWriteHeaderTwice() {
+        final File bcfOutputFile = VariantBaseTest.createTempFile("testWriteAndReadVCF.", ".bcf");
+
+        Tribble.indexFile(bcfOutputFile).deleteOnExit();
+        Set<VCFHeaderLine> metaData = new HashSet<VCFHeaderLine>();
+        Set<String> additionalColumns = new HashSet<String>();
+        final SAMSequenceDictionary sequenceDict = createArtificialSequenceDictionary();
+        final VCFHeader header = createFakeHeader(metaData, additionalColumns, sequenceDict);
+
+        // prevent writing header twice
+        final VariantContextWriter writer1 = new VariantContextWriterBuilder()
+                .setOutputFile(bcfOutputFile).setReferenceDictionary(sequenceDict)
+                .setOptions(EnumSet.of(Options.INDEX_ON_THE_FLY))
+                .build();
+        writer1.writeHeader(header);
+        writer1.writeHeader(header);
+        writer1.add(createVC(header));
+        writer1.close();
+    }
+
+    @Test(expectedExceptions = IllegalStateException.class)
+    public void testChangeHeaderAfterWritingHeader() {
+        final File bcfOutputFile = VariantBaseTest.createTempFile("testWriteAndReadVCF.", ".bcf");
+
+        Tribble.indexFile(bcfOutputFile).deleteOnExit();
+        Set<VCFHeaderLine> metaData = new HashSet<VCFHeaderLine>();
+        Set<String> additionalColumns = new HashSet<String>();
+        final SAMSequenceDictionary sequenceDict = createArtificialSequenceDictionary();
+        final VCFHeader header = createFakeHeader(metaData, additionalColumns, sequenceDict);
+
+        // prevent changing header if it's already written
+        final VariantContextWriter writer2 = new VariantContextWriterBuilder()
+                .setOutputFile(bcfOutputFile).setReferenceDictionary(sequenceDict)
+                .setOptions(EnumSet.of(Options.INDEX_ON_THE_FLY))
```

Comment:
Removed INDEX_ON_THE_FLY option. But sequence dictionary has to be set because INDEX_ON_THE_FLY is a default option anyway, and without setting the sequence dictionary, writer cannot be created (an exception gets thrown).